### PR TITLE
Machine Learning Optimization Template

### PR DESCRIPTION
**Pull request by Wen-Tian-Pineapple**

### CODE REVIEW

**Comment by msaligane** [comment]
@gkielian Do you think you can take a look?

File: openfasoc/MLoptimization/image2.png
Comment (gkielian) [inline openfasoc/MLoptimization/image2.png:1]: Cool, please change the name from image2.png to mean_reward_vs_environemnet_step.png or someting human readable.

File: openfasoc/MLoptimization/eval.py
Comment (gkielian) [inline openfasoc/MLoptimization/eval.py:111]: Remove commented code if unused.

File: openfasoc/MLoptimization/gen_spec.py
Comment (gkielian) [inline openfasoc/MLoptimization/gen_spec.py:15]: Add variables for each of these constants, and provide descriptive names for each.

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:11]: remove unused #

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:13]: Refactor file to if name main format and all argparse's into a function called parse_args

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:31]: Provide arg parse params for these with defaults being the current settings above.

Also provide descriptions for each param via argparse to help the user understand what each does.

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:29]: Is this for parallelization? Would this depend on the cores of the machine being run on? If so should be parameterized via argparse for sure.

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:33]: results should not be saved to an absolute path, instead save them somewhere in this project, and add that directory to your `.gitignore`

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:37]: Stick to snake case for the name here

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:36]: Then this should be a parameter set via argparse.

File: openfasoc/MLoptimization/model.py
Comment (gkielian) [inline openfasoc/MLoptimization/model.py:45]: Remove absolute path, also worried restoring this timestamped file wouldn't work on a new user's machine

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:24]: Remove unused #

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:72]: remove, unused

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:87]: If too cumbersome to place as argparse commands, should at least be constants with meaningful names, e.g. MIM_CAP_ROW_MIN, MIM_CAP_ROW_MAX, MIM_CAP_ROW_STEP.

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:95]: remove unused code

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:103]: remove unused code

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:151]: remove unused code

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:165]: what does it return?

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:173]: remove unused code

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:180]: remove unused code

File: openfasoc/MLoptimization/run_training.py
Comment (gkielian) [inline openfasoc/MLoptimization/run_training.py:258]: Remove unused code

**Comment by msaligane** [comment]
@alibillalhammoud can you please test this out?

File: openfasoc/MLoptimization/eval.py
Comment (andylithia) [inline openfasoc/MLoptimization/eval.py:8]: Directory names have changed.
At this point the ML module is a one-off business, and the API is not well established (compared to GLayout), in the long run this kind of project-specific hack should be avoided.

File: openfasoc/MLoptimization/eval.py
Comment (andylithia) [reply openfasoc/MLoptimization/eval.py:65]: Also would be nice to have comments

File: openfasoc/MLoptimization/model.py
Comment (andylithia) [reply openfasoc/MLoptimization/model.py:11]: Same comments as before:
Directory names have changed.
At this point the ML module is a one-off business, and the API is not well established (compared to GLayout), in the long run this kind of project-specific hack should be avoided.

File: openfasoc/MLoptimization/run_training.py
Comment (andylithia) [reply openfasoc/MLoptimization/run_training.py:24]: Same comments as before:
Directory names have changed.
At this point the ML module is a one-off business, and the API is not well established (compared to GLayout), in the long run this kind of project-specific hack should be avoided.

File: openfasoc/MLoptimization/sample_spec.yaml
Comment (andylithia) [reply openfasoc/MLoptimization/sample_spec.yaml:100]: I think it's a better idea to provide the script that generates this yaml file instead of uploading raw data like this?
`gain_min` is from one of the earlier runs, I think it's deprecated and need to be updated.
`FOM` is also very old, need to be updated.

**Review by andylithia — APPROVED** [review]
Need a refactor at some point. For now I'll merge this, and work in the new branch 'RL' to address all the issues

**Comment by msaligane** [comment]
@andylithia please do not merge things without my approval.

**Pull request merged**
